Fix bug when source is already merged in dest

mdbooth opened the pull request:
If:
* source has no new changes
* changes are already merged in dest

We would try to open a new PR containing no new commits. This resulted
in a 422 response from GitHub.

With this change we explicitly check if a PR is required.

Commit:
Fix bug when source is already merged in dest

If:
* source has no new changes
* changes are already merged in dest

We would try to open a new PR containing no new commits. This resulted
in a 422 response from GitHub.

With this change we explicitly check if a PR is required.

mandre commented on src/merge_bot/merge_bot.py:
```diff
@@ -57,54 +57,63 @@ def configure_commit_info(repo, bot_name, bot_email):
-                gitwd.remotes.source.refs[source.branch].commit,
-            ),
-        )
-        return True
+    do_merge()
 
-    if gitwd.active_branch.commit != orig_commit:
-        logging.info("Destination can be fast-forwarded")
-        return True
+    # Check if we've already merged
+    if git_is_ancestor(gitwd, f"merge/{merge.branch}", f"dest/{dest.branch}"):
+        logging.info("Destination is already up-to-date")
+        return False
 
-    logging.info("No merge is necessary")
-    return False
+    logging.info("Destination needs to be updated")
+    return True
```
So this is really the only interesting bits here, the rest is only refactoring IIUC. You could have made it clearer if you had split the bugfix and refactoring into separate commits.

mandre commented on src/merge_bot/merge_bot.py:
```diff
@@ -57,54 +57,63 @@ def configure_commit_info(repo, bot_name, bot_email):
-    if merge.branch in gitwd.remotes.merge.refs:
-        # Check if we have already pushed a merge to the merge branch which
-        # contains the current head of the source branch
+    def do_merge():
```
Interesting. Why do you wrap this code in a nested function and immediately call it after? Is it to better convey the meaning of this code? Not advocating against the change itself, but it doesn't belong to this commit as it makes it harder to review the functional change.

mdbooth replied:
It's syntactic sugar when you want to shortcut out of a conditional block. It's much simpler to return than to nest a dozen ifs.

mandre replied:
This it what nested functions are generally used for, but this isn't what you're doing here, are you?
Since you're calling `do_merge()` immediately after defining the inner function, and only here, you're effectively only adding one indentation level for this code. Hence my question about the purpose of this change.

mdbooth commented:
Thanks @mandre. I've converted this to draft until I've addressed your comments. This bug results in a bogus error message but is otherwise harmless, so it's not worth pushing it through. I also feel like we're beyond the point of needing more tests.

mandre commented:
> Thanks @mandre. I've converted this to draft until I've addressed your comments. This bug results in a bogus error message but is otherwise harmless, so it's not worth pushing it through. I also feel like we're beyond the point of needing more tests.

Good point about tests. It's going to be important if we want wider adoption of merge-bot inside the org and beyond.

dulek commented:
@mdbooth: I attempted to fix the same in #13. Feel free to check out my approach there.

mandre commented:
This is no longer needed
/close
